python/ray/util/sgd/torch/training_operator.py

```diff
@@ -109,13 +153,233 @@ def _set_timers(self, timers):
+        self._original_models = models
+        if not isinstance(self._original_models, Iterable):
+            self._original_models = [self._original_models]
+        assert all(
+            isinstance(model, nn.Module) for model in self._original_models), (
+                f"All models must be PyTorch models: {self._original_models}.")
+        if self.use_gpu and torch.cuda.is_available():
+            self._original_models = [
+                model.cuda() for model in self._original_models
+            ]
+
+        logger.debug("Registering optimizers.")
+        self._optimizers = optimizers
+        if not isinstance(self._optimizers, Iterable):
+            self._optimizers = [self._optimizers]
+
+        if schedulers:
+            logger.debug("Registering scheduler.")
+            self._schedulers = schedulers
+            if not isinstance(self._schedulers, Iterable):
+                self._schedulers = [self._schedulers]
+        else:
+            self._schedulers = None
+
+        if criterion:
+            logger.debug("Registering loss.")
+            self._criterion = criterion
+            if self.use_gpu and torch.cuda.is_available():
+                if hasattr(self._criterion, "cuda"):
+                    self._criterion = self._criterion.cuda()
+        else:
+            self._criterion = None
+
+        logger.debug("Setting up Apex.")
+        if self.use_fp16 and amp:
+            self._models, self._optimizers = amp.initialize(
+                self._models, self._optimizers, **self._apex_args)
+            self._amp = amp
+
+        if self._wrap_ddp:
+            logging.debug("Setting up DDP for models.")
+            self._models = [
+                DistributedDataParallel(model, device_ids=self.device_ids)
+                for model in self._original_models
+            ]
+        else:
+            self._models = self._original_models
+
+        if len(self._models) == 1:
+            return_vals.append(self._models[0])
+        else:
+            return_vals.append(self._models)
+
+        if len(self._optimizers) == 1:
+            return_vals.append(self._optimizers[0])
+        else:
+            return_vals.append(self._optimizers)
+
+        if self._criterion is not None:
+            return_vals.append(self._criterion)
+
+        if self._schedulers is not None:
+            if self.scheduler_step_freq is None:
+                raise ValueError("scheduler_step_freq passed into "
+                                 "TorchTrainer cannot be None if you "
+                                 "are registering schedulers. Set this to "
+                                 "'manual' if you will be manually stepping "
+                                 "the schedulers.")
+            if len(self._schedulers) == 1:
+                return_vals.append(self._schedulers[0])
+            else:
+                return_vals.append(self._schedulers)
+
+        return tuple(return_vals)
+
```

Comment:
Thanks for the feedback!

For the first point, would you say the name of the method is confusing or the examples in the docs/docstrings?

For the second point, we thought it would be better for the user to explicitly set these attributes so they wouldn't have to access any hidden variables if providing custom training/validation. These variable names don't necessarily have to be `self.model`, `self.optimizer`, etc. the user can set it to whatever they want like `self.classifier`.

Reply:
1) I think it's just the name of the method, but I don't feel strongly about this.

2) Fine with me, let's leave it as is.